### ISSUE TITLE
Clarify security philosophy around update & delete

### ISSUE THREAD

**Issue by lavalamp** [kubernetes/kubernetes#24891]
We had some vigorous discussion in the API machinery SIG earlier today about what a system administrator should expect when granting a user update rights on a given object or class of objects.

My own position is that update is strictly more powerful than delete and you shouldn't give anyone update powers over an object you don't want them to do destructive things to.

The opposing position is, as @derekwaynecarr put it, update : delete :: paint house : burn house down.

We discussed whether it's reasonable to expect system admins to understand the exact power update gives you over any particular sort of object; some objects allow more destructive updates than others. We discussed extracting "harmless" field updates into subresources which would be safe, with source code that is easily checkably safe.

We discussed how our current update validation path appears to be a fertile source of CVEs, and whether or not we should expect that to change in the future.

We didn't come to clear conclusions on these issues, and whatever we do decide the system's philosophy should be here, we need committers and reviewers to understand it. So I'm opening this issue so we can hopefully come to some agreement.

The issue that triggered this is the garbage collector: #23656 and also see discussion on its API PR: #23928

**Comment by lavalamp** [comment]
Tagging discussion participants in case I have misrepresented views. @smarterclayton @erictune @liggitt @deads2k @caesarxuchao @nikhiljindal

**Comment by smarterclayton** [comment]
Other points:

1. We want to define "roles" that are logically consistent to admins and users, and within those roles we are relatively safe (we did this as part of the original authz design, and need to revisit)
2. Improve the process whereby we design API objects for security with the right checks and balances - being config driven (PUT / CREATE work cleanly) and mutable often conflicts with normal security policy (setting externalIP with a custom network plugin can end in tears)
3. We have danced around field level authorization many times - in many respects it is like a generic defaulter or generic limit ranger, and at some point we need to decide on the investment.

**Comment by thockin** [comment]
Maybe the delete-related fields should not be updateable?

I'd be interested to see more finer-grained subresource "methods" as the
way of delegating control over various things.

**Comment by smarterclayton** [comment]
Sub resources unfortunately don't solve the "is CREATE safe for this" -
either you end up not giving CREATE to end users (because you want to set
those fields), or you can't set those fields on CREATE (in which case they
aren't sub resources, they are separate resources).  Since create is the
most common verb, I don't think it's a real solution except for system
actors that only react to objects, not objects that are set by end users.

I.e., service.spec.externalIPs and pod.spec.nodeName both allow users to
potentially abuse a system (nodeName allows you to steer pods to nodes and
deny service, externalIPs can be used to impersonate hosts).  We would have
to prevent them from being set on CREATE to be safe.

**Comment by thockin** [comment]
I was mostly thinking of update - a lot of fields are sane to set but
awkward to update (or want to be access-controlled differently)  Creating
an object with a deletion time is valid (albeit weird, maybe).  Updating
the deletion time is a somewhat distinct action from updating the object
and unlikely to be done in tandem (I think).  So gating deletion by a
subresource seems 100% reasonable.  We should probably have /label and
/annotate subresources on every object, too, though that seems sane in a
plain update, as well.

**Comment by smarterclayton** [comment]
We should distinguish between fields that are safe within a resource
(creator is only hurting themselves) and safe outside a resource (field
allows creator to impact others).  The former is fine for create w/ update
subresources, the latter is not.

Another challenge to subresources - if a field is not dangerous until
someone starts using it in a piece of automation outside the core system,
and the only way to make it safe is to add a subresource, then the layered
controller must either get core Kube to change to have a subresource in
order to be safe, or create a virtual proxy resource that guards the core
resource but writes data to another object.  Ie any authoritative decision
data has to be in an object that has exactly the scope of the edits.
Annotations don't cut it unless they are signed by the controller.
Example: externalIPs should be a different resource.

On Apr 30, 2016, at 7:52 PM, Tim Hockin <notifications@github.com> wrote:

I was mostly thinking of update - a lot of fields are sane to set but
awkward to update (or want to be access-controlled differently) Creating
an object with a deletion time is valid (albeit weird, maybe). Updating
the deletion time is a somewhat distinct action from updating the object
and unlikely to be done in tandem (I think). So gating deletion by a
subresource seems 100% reasonable. We should probably have /label and
/annotate subresources on every object, too, though that seems sane in a
plain update, as well.

On Thu, Apr 28, 2016 at 7:36 PM, Clayton Coleman <notifications@github.com>
wrote:

> Sub resources unfortunately don't solve the "is CREATE safe for this" -
> either you end up not giving CREATE to end users (because you want to set
> those fields), or you can't set those fields on CREATE (in which case they
> aren't sub resources, they are separate resources). Since create is the
> most common verb, I don't think it's a real solution except for system
> actors that only react to objects, not objects that are set by end users.
>
> I.e., service.spec.externalIPs and pod.spec.nodeName both allow users to
> potentially abuse a system (nodeName allows you to steer pods to nodes and
> deny service, externalIPs can be used to impersonate hosts). We would have
> to prevent them from being set on CREATE to be safe.
>
> On Thu, Apr 28, 2016 at 1:24 AM, Tim Hockin <notifications@github.com>
> wrote:
>
> > Maybe the delete-related fields should not be updateable?
> >
> > I'd be interested to see more finer-grained subresource "methods" as the
> > way of delegating control over various things.
> >
> > On Wed, Apr 27, 2016 at 4:03 PM, Clayton Coleman <
> notifications@github.com
> > >
> > wrote:
> >
> > > Other points:
> > >
> > > 1. We want to define "roles" that are logically consistent to admins
> > > and users, and within those roles we are relatively safe (we did this
> as
> > > part of the original authz design, and need to revisit)
> > > 2. Improve the process whereby we design API objects for security with
> > > the right checks and balances - being config driven (PUT / CREATE work
> > > cleanly) and mutable often conflicts with normal security policy
> (setting
> > > externalIP with a custom network plugin can end in tears)
> > > 3. We have danced around field level authorization many times - in
> > > many respects it is like a generic defaulter or generic limit ranger,
> and
> > > at some point we need to decide on the investment.
> > >
> > > —
> > > You are receiving this because you are subscribed to this thread.
> > > Reply to this email directly or view it on GitHub
> > > <
> >
>
https://github.com/kubernetes/kubernetes/issues/24891#issuecomment-215258657
> > >
> >
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly or view it on GitHub
> > <
>
https://github.com/kubernetes/kubernetes/issues/24891#issuecomment-215313220
> >
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <
https://github.com/kubernetes/kubernetes/issues/24891#issuecomment-215613934
>
>

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/24891#issuecomment-216003493>

**Comment by thockin** [comment]
I don't know how to parse this.  I can't tell exactly what you're
arguing for (or against).

On Sat, Apr 30, 2016 at 5:31 PM, Clayton Coleman
<notifications@github.com> wrote:
>
> We should distinguish between fields that are safe within a resource
> (creator is only hurting themselves) and safe outside a resource (field
> allows creator to impact others). The former is fine for create w/ update
> subresources, the latter is not.

It's not even about what is "safe" in my mind, it's whether things make sense.

> Another challenge to subresources - if a field is not dangerous until
> someone starts using it in a piece of automation outside the core system,
> and the only way to make it safe is to add a subresource, then the layered
> controller must either get core Kube to change to have a subresource in

core Kube as in "upstream" ?  That's all I am talking about.  And
anyway it's a little late for this discussion wrt v1, but if/when we
do v2 I want to push for more "method-like" subresources.

> order to be safe, or create a virtual proxy resource that guards the core
> resource but writes data to another object. Ie any authoritative decision
> data has to be in an object that has exactly the scope of the edits.
> Annotations don't cut it unless they are signed by the controller.
> Example: externalIPs should be a different resource.
>
>

**Comment by derekwaynecarr** [comment]
I am glad it was so easy to summarize my position!

Another issue is that as sub resources proliferate things like kubectl edit
or apply may become less useful?

**Comment by bgrant0607** [comment]
@derekwaynecarr @smarterclayton 

My analogy would be: Delete : Update :: Suicide : Mass Murder.

Update is often more powerful than delete. For instance, the ability to update a pod template provides the power to run anything with the credentials granted to that controller's pods. Or, updating a service could expose confidential information.

**Comment by derekwaynecarr** [comment]
I agree that that is the case some of the time but not all of the time.
For a generic API server, I don't think we can extend that claim
universally.

**Comment by derekwaynecarr** [comment]
I am a few days removed from this discussion.  To me, the pressing issue is
if anything is broken with this change going forward.  I thought @liggitt
enumerated a set of good examples.

**Comment by smarterclayton** [comment]
From a risk perspective, delete is usually more straightforward to
understand the implications of than update... until this change :)  So
update is much more dangerous in terms of the author's ability to reason
about the possible outcomes.  With this change, delete could propagate
throughout the system and the deleter has no way of knowing what the impact
is.  If we don't add additional semantics though, delete is still simpler
to reason about.

**Comment by thockin** [comment]
Yeah, nothing here has made me believe that update-and-set-deletion-time is
actually what we WANT to support.

**Comment by techtonik** [comment]
Nothing explains security issues better than attack vectors. So..

**Comment by sftim** [comment]
I can think of a bunch of different CRDs / extension APIs where being able to **delete** is obviously more problematic than being allowed to **update**.

What those have in common is that there are immutable fields, and an **update** can't change those, or perhaps the underlying actual system is somehow immutable. 

- if you are allowed to **delete** and **create**, then you can make changes that an **update** might not allow
- if you are allowed to **delete** something from a global namespace, you or an accomplice can then make a replacement in the global namespace with a different owner / different access policy / whatever
  - and maybe also trigger a TOCTOU risk, depending on circumstances

**Comment by lmktfy** [comment]
How do we feel about this now that ValidatingAdmissionPolicy is stable and _right there_ if you want to restrict changes (including deletes)?
